### PR TITLE
Support function in custom node's default argument expression and tool tip enhancement

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

This PR is to support function in custom nodes' default argument expression, and enhances the tooltip of custom nodes' input port:
  * Previously, only primitive types are allowed, say `x : int = 21`. Now complex expression is supported: 

  * Allow comments in input node. Tooltip of the input port of custom node also is enhanced to display default value expression *and* comments:

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.

### Reviewers

@Benglin  PTAL.

### FYIs

@Benglin , definitely there are some code cleanup work around the parameter editor. I already created a class `CodeComletionEditor`, which finally should be merged with `CodeBlockEditor`. As the changes in this PR is not small, I'll do that cleanup in other PR.

### CODE REVIEW

**Comment by ikeough** [comment]
@ke-yu This is awesome!

File: src/DynamoCoreWpf/Views/CodeBlocks/CodeBlockEditor.xaml.cs [hunk 1]
```diff
@@ -145,66 +145,6 @@ public string Code
         
         #endregion
 
-        #region Syntax highlighting helper methods
```
Comment (ke-yu) [inline src/DynamoCoreWpf/Views/CodeBlocks/CodeBlockEditor.xaml.cs]: They are moved to `CodeHighlightingRuleFactory`.

**Comment by ikeough** [comment]
@ke-yu I hadn't realized you added comments as well. This just keeps getting better :smile:

**Comment by ke-yu** [comment]
@ikeough , haha, it is not an ideal solution, but I think it probably would be helpful.

File: src/DynamoCore/Graph/Nodes/CustomNodes/Function.cs [hunk 2]
```diff
@@ -381,9 +383,14 @@ protected override void DeserializeCore(XmlElement nodeElement, SaveContext cont
             var parseParam = new ParseParam(this.GUID, parseString, resolver);
 
             if (EngineController.CompilationServices.PreCompileCodeBlock(ref parseParam) &&
-                parseParam.ParsedNodes != null &&
                 parseParam.ParsedNodes.Any())
             {
+                var parsedComments = parseParam.ParsedComments;
+                if (parsedComments != null && parsedComments.Any())
```
Comment (Benglin) [inline src/DynamoCore/Graph/Nodes/CustomNodes/Function.cs]: No need to check for `null`

File: src/DynamoCore/Graph/Nodes/CustomNodes/Function.cs [hunk 3]
```diff
@@ -393,6 +400,11 @@ protected override void DeserializeCore(XmlElement nodeElement, SaveContext cont
                     if (inputSymbol.Contains('='))
                         defaultValue = node.RightNode;
 
+                    if (parseParam.Errors != null && parseParam.Errors.Any())
+                        this.Error(parseParam.Errors.First().Message);
+                    else if (parseParam.Warnings != null && parseParam.Warnings.Any())
+                        this.Warning(parseParam.Warnings.First().Message);
+
```
Comment (Benglin) [inline src/DynamoCore/Graph/Nodes/CustomNodes/Function.cs]: Optional `null` checks?

**Comment by dimven** [comment]
Isn't this already partially supported? I've been using this for a while in 0.82:

**Comment by ke-yu** [comment]
@dimven , weird! I don't think I added the support for function call in 0.8.2. Just tested, I'm even not able to type '(' (but yeah, you could modify.dyn file to bypass some UI limitation in 0.8.2)

**Comment by dimven** [comment]
You can't type, but you can paste directly in the input node's input ;)

**Comment by ke-yu** [comment]
@dimven  ouch! :-) anyway, with this one you'd get auto-completion.

**Comment by dimven** [comment]
:+1:

**Comment by mostaphaRoudsari** [comment]
Hi @ke-yu @ikeough! This is very useful. any plans to do something similar for outputs so one can add descriptions to outputs? Or maybe it is already possible and I don't know about it. Let me know if you want me to open a new issue for this

**Comment by ke-yu** [comment]
Hi @mostaphaRoudsari , no plan for output yet. Do you think it would be useful? 

Include @riteshchandawar @Racel

**Comment by mostaphaRoudsari** [comment]
Hi @ke-yu! Yes. It will be very helpful for the nodes that I'm developing. It will make outputs understandable and in some cases it tells the user what they can do with the output. Here is an example:

image

For now I made an extra output to have description for all the outputs which doesn't look to be that useful.

image

Is this something that takes so much work to add?

**Comment by mostaphaRoudsari** [comment]
Also in the input name it removes _ at the end of the  name. so *\_analysisPeriod\_* is changed to *_analysisPeriod*

**Comment by Racel** [comment]
@mostaphaRoudsari @ke-yu - FYI -I have added this to our backlog.

**Comment by mostaphaRoudsari** [comment]
Thanks @Racel!

@ke-yu while we are talking about the inputs/outputs I want also to add this which looks like a bug. If I add description to an input without adding a type it won't work as expected.

image

image

I'm using Dyamo 0.9.1.3360

**Comment by ke-yu** [comment]
@mostaphaRoudsari it is a bug. @monikaprabhu

**Comment by ParametricMonkey** [comment]
@ke-yu Can you tell me why these inputs would be failing? It worked on v0.9.1 but not 0.9.2

error

http://dynamobim.org/forums/topic/custom-node-failing-in-v0-9-2/

**Comment by kronz** [comment]
@ParametricMonkey Can you try this in 1.0?  looks like it works for me

**Comment by ParametricMonkey** [comment]
@kronz yes that worked. so 'string' has to be lower case now and 'FilePath' is replaces with 'string'? Are there any other input changes that I should be aware of and where can I find the documentation of it? Interestingly Ryhnao and Lunchbox worked fine in v1.0.0 when i tested this.
Thanks

**Comment by kronz** [comment]
@ParametricMonkey There aren't any changes that I know of to this behavior.  I believe the only changes we have made here in 1.0 are to allow //comments to be added to output ports as well.  I have added a task to our Dynamo Primer backlog to expand on the section that demonstrates custom inputs and defaults for custom nodes.

**Pull request merged**